text/0000-global-panic-handler.md

```diff
@@ -0,0 +1,183 @@
+- Feature Name: panic_handler
+- Start Date: 2015-10-08
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+When a thread panics in Rust, the unwinding runtime currently prints a message
+to standard error containing the panic argument as well as the filename and
+line number corresponding to the location from which the panic originated.
+This RFC proposes a mechanism to allow user code to replace this logic with
+custom handlers that will run before unwinding begins.
+
+# Motivation
+
+The default behavior is not always ideal for all programs:
+
+* Programs with command line interfaces do not want their output polluted by
+  random panic messages.
+* Programs using a logging framework may want panic messages to be routed into
+  that system so that they can be processed like other events.
+* Programs with graphical user interfaces may not have standard error attached
+  at all and want to be notified of thread panics to potentially display an
+  internal error dialog to the user.
+
+The standard library [previously
+supported](https://doc.rust-lang.org/1.3.0/std/rt/unwind/fn.register.html) (in
+unstable code) the registration of a set of panic handlers. This API had
+several issues:
+
+* The system supported a fixed but unspecified number of handlers, and a
+  handler could never be unregistered once added.
+* The callbacks were raw function pointers rather than closures.
+* Handlers would be invoked on nested panics, which would result in a stack
+  overflow if a handler itself panicked.
+* The callbacks were specified to take the panic message, file name and line
+  number directly. This would prevent us from adding more functionality in
+  the future, such as access to backtrace information. In addition, the
+  presence of file names and line numbers for all panics causes some amount of
+  binary bloat and we may want to add some avenue to allow for the omission of
+  those values in the future.
+
+# Detailed design
+
+A new module, `std::panic`, will be created with a panic handling API:
+
+```rust
+/// Unregisters the current panic handler, returning it.
+///
+/// If no custom handler is registered, the default handler will be returned.
+///
+/// # Panics
+///
+/// Panics if called from a panicking thread. Note that this will be a nested
+/// panic and therefore abort the process.
```

Comment:
When a panic handler is invoked, perhaps it could be considered as "being taken"? That would mean that this function would never need to panic as if called while panicking it'd just continue to return the default handler.